### PR TITLE
`test` to `[[` where available

### LINKED ISSUE #265: Bug: `ask` prompt is not clearing correctly, causing ask prompt and answer to duplicate and cursor to move in-between data
Image

### LINKED ISSUE #268: Suggestion: instead of requiring `sudo` for `is-present` and `is-missing` failures, test if assumptions are correct
Currently my assumption with `is-present` and `is-missing` in the upcoming #261 is that if they fail to detect existence, it could just be that we don't have the permissions to determine it - however I haven't actually tested this yet, and there are tests for it - could be that a different exit status is returned if it does exist but we don't have permissions - however I'd imagine that would be a security issue, as one could brute-force the structure of a filesystem they don't have access to.

If the exit status of missing, and no access, are the same, then one alternative would be to `touch` at that location, and if we can do that, then it does mean that the path didn't actually exist. Note that this will only work for paths which parents do exist, and we can't do `mkdir -p` for this, as it would create a heirarchy of many levels of non-existence which is unexpected behaviour.

Asking GitHub Copilot about this was useless.

### PR DESCRIPTION
**Pull request by balupton**
/close #265 #268

Using `[[ ... ]]` instead of `test` is a convention change that improves performance and resolves edge cases where `test -n "$a" -a "$b" = "$c"` fails when `a='>'`.

This PR also features these highlights:

- clarified help text around return statuses, and help text will now mention companion commands
- `ask`: fixed clearing issue /close #265
- reduce sudo escalations by introduction of auto-escalating `is-*` filesystem commands for filesystem interactions which use a new `is-accessible` command to determine if a sudo escalation is needed, benefiting #233 and closes #268, full list of commands:
    - `is-accessible`
    - `is-broken-symlink`
    - `is-directory`
    - `is-empty-directory`
    - `is-empty-file`
    - `is-executable`
    - `is-file`
    - `is-missing`
    - `is-nonempty-file`
    - `is-not-directory`
    - `is-not-symlink`
    - `is-owner`
    - `is-present`
    - `is-readable`
    - `is-symlink`
    - `is-writable`
    - `echo-if-directory`
    - `echo-if-executable`
    - `echo-if-file`
    - `echo-if-present`
    - `fs-absolute`
    - `fs-own`
    - `fs-dequarantine`
    - `fs-parents`
    - `fs-realpath`
    - `fs-structure`
    - `gocryptfs-helper`
    - `secret`
- `fs-rm`:  completely rewritten for a superior experience, with revised functionality to simplify and secure its human and programmatic usage (such as in `setup-dns`), also supports `--trash`
- `fs-structure`: rewritten, and now used in `fs-own`, `fs-rm`, `fs-trim`, `gocryptfs-helper`, `is-same`, `setup-util`, `eject-all`
- added new `fs-trim` to assist with `fs-rm` under certain circumstances
- `choose`:
    - properly support cancelling to no selection when not required
    - support `--truncate-body` for when we allow the question's body/description/details to be truncated to make way for items, used by `fs-rm` and `fs-trim`, defaults to disabled
    - support `--default-all` for defaulting to selection of everything, defaults to disabled
    - support `--confirm-cancel` which now defaults to enabled
    - use `styles.bash` values for terminal title and resize modifications
    - fix typos in the legend
    - add support for the `r` key for reversion, along with the prior `z` key
- `dorothy`:
    - fix `dnf` ecosystem changes breaking installation on dnf setups
    - `install` action: cleaner removal `.profile` modification if no `bash`, `dash`, `ksh`
    - `lint` and `test` actions now sort files by most recent first, and `test` action will call the new `git-helper umt` command and action on CI to ensure clone dates are correct
    - new `commands` action: lists all the available commands, recommended now in post-install instructions, uses new `echo-if-present` command, however doesn't yet use `fs-structure`
    - `todos` action: support checking core and beta commands, correct detection of usage, correct error status reporting, clearer indication when help text is missing, fix missing and incorrect local vars
    - use `is-ci` instead of an archaic technique
    - `permissions` action: no longer trim junk files, as they are already git-ignored
- added a new `sources/tests.bash` for assisting the fs commands with their tests
- added a new `is-owner` command for detecting if a user or group has ownership of a path
- added a new `is-root` command for detecting if a user is a root user (maybe one day this should support the root group as well)
- `read-key`: added a `--continue` option for debugging, but most importantly reduced the subsequent timeout from `0.0001` to `0.01` to prevent situations where a key was read only partially and caused incorrect keys to be assumed, causing say `ask`, `choose`, `confirm` to interpret an escape key that wasn't actually pressed
- `setup-util`:
    - load `styles.bash` instead of calling `echo-style` for performance
    - use the new `fs-rm` flow for removing leftover and conflicting paths
    - `rm_helper_confirm` to `rm_helper_which` to better explain its purpose
    - `echo_fallback` to `log_fallback`
    - improve styling of `APT_REPO` `http` vs `https` warning
    - use the new `fs-structure` for outputting when a download build glob fails
    - ensure all output goes not to stdout but to the terminal device file, such that `$(... something that calls setup-util ...)`` does not get a bunch of install junk in the result, as was the case for `tests.bash` calling of `symlink-helper` which installed coreutils for realpath
        - this was mostly achieved by adding the missing `outputpipe="$terminal_device_file"` to when `use_alt_screen_buffer` was `no`
    - also ensure errors go to stderr where appropriate
- `sudo-helper`:  add alternative values for `--sudo` option, to enable optional escalation under various circumstances, and simplified logic
- `symlink-helper`: support `--target=` as alias for `--existing=`, and improve `--quiet` and `--verbose` handling to ensure outputs match expectations, and with nothing going to stdout

Other changes to the core:

- everywhere:
    - remove unnecessary `--status=0` option with `eval-tester`
    - source code is now spell checked
    - use `=~ ^(...|...)$` instead of multiple `== ... ||` statements for easier human parsing
- ci:
    - test on ubuntu arm runners
    - compat with newer dnf versions
    - upgrade lint tooling
 - `init.sh`: fix a missing `-d`
- `config-edit`: `--editer` changed to correct spelling of `--editor`
- `command-working`: add `rmtrash` to the exception
- `down`:
    - added `--progress` option that is enabled by default, except on CI where it is disabled by default which prevents progress bars from appearing on CI
    - fix an issue where the temporary download directory might not exist
- `edit`: don't output `None of the configured editors were available` if there are no configured editors, such as the case on a new machine (review this later, as we may still want a warning, but a better warning on more specific certain situations)
- `echo-style`: fixed empty `--=` and `--` usage, the implementation seemed broken from the start
- `eval-helper`:
     - categorise `sudo-helper` automatically as a shapeshifter, as it is impossible to know how many lines it used while attempting passwords, so impossible to clear it accurately in non-shapeshifter mode
     - all messages now correctly go to the terminal device file, such that stdout captures are not corrupted with messages
     - if using `echo-revolving-door` don't output anything to stdout, as anything would not be correct as it would be truncated
- `flush-dns`: use `--shapeshifter` on the `sudo-helper` calls for interacting with dns responders to ensure it is all rendered correctly
- `fs-dequarantine`: use `/usr/bin/xattr` as homebrew can cause conflicts with `xattr`
- `fs-filename`: fix help convention
- `fs-own`:
    - accept `--recursive` with variant spelling
    - add `--root` for setting ownership as the platform's root user and group
    - add support for `--verbose` on macos
 - `fs-temp`: clarified `--touch` behaviour, and support `--no-touch` on directories
- `get-flag-value`: implement `--fallback-on-empty` option used by `fs-absolute`
- `get-json-api-url.awk`: fix typo in var name
- `get-profile`: correct spelling of `possesive-pronoun` to `possessive-pronoun`
- `get-terminal-color-support`: clarified help text on return statuses in different modes
- `get-terminal-theme`: use `is-integer` sanity check for `colorfgbg` source
- `git-helper`: add new `umt` action to apply commit modification times to cloned git repos, used by `dorothy test` on CI
- `github-download`: fix a missing `-n` on `$pathname` exclusivity check
- `is-empty-value`: added tests and made performant
- `is-admin`: fix multiple user handling
- `is-even`: add tests, and only support integers
- `is-generic`: added empty string as generic, added more tests
- `is-missing`: fix incorrect help text on exit status
- `is-needle`: added more tests, added `--all` and `--any` option modifiers
- `is-odd`: add tests, and only support integers
- `mount-helper`: fix two missing `-n`
    - samba username and password check: `if test -n "$option_username" -o "$option_password"; then` would have caused an error state if only password is defined
    - `if test -n "$temp" -a "$option_label"; then` changed to just `if [[ -n $temp ]]; then` as in this case, it seems better to not exist, as we always want the first line of `$temp` regardless of `$option_label` (however, then it is more a sanity procedure, but a necessary one)
    - fixed detection of forced unmount
- `researchgate-rename`: fix bad `is-same` call
- `setup-dns`: use new `fs-rm` flow for removing unwanted configuration and installation files, and fix potential regression issue when removals were already removed
- `setup-git`: fix a missing `test` / `[[`
    - `if test "$GPG_SIGNING_AGENT" != 'op' && "$GPG_SIGNING_AGENT" != 'krypton'; then` affected `GPG_SIGNING_AGENT` when not using 1password
- `setup-mac-brew`: don't enforce sudo for the cleanup `fs-rm`, as `fs-rm` is now intelligent enough to handle escalation itself if needed
- add `setup-util-asciinema`
- `setup-util-eza`: use `https` for `APT_REPO`
- add `setup-util-firefox`
- add `setup-util-qpdf` used by `pdf-decrypt`
- add `setup-util-tor-browser`
- `setup-util-trash`: updated for a library that actually works, and uninstall the ones that don't
- `setup-util-xcode`: use shapeshifter for the xcode license agreement, as it requires it as it modifies cursor
- `waiter`: fix missing `-n` on `--stdout`, `--stderr`, and `--tty` option handling
- `config/dns.bash`: removed outdated configuration options
- `config/styles.bash`: fixed terminal resize and end styles not being found due to typo
- `docs/errors.md`: detailed the errors that dorothy actually uses
- `sources/nvm.sh`: fix for edge case where nvm fails to detect system `node`

deprecated core commands:

- `echo-segment` replaced by `echo-style`, this was done a year or more ago, before we had the deprecated directory
- `fs-size` replaced by `fs-structure` or `get-size` depending on your use case
- `get-array-*` and `is-array-*` commands replaced by `echo-trim-empty-lines` yes that is correct and yes it is weird, but those commands were weird
- `is-array-(empty|full|partial)` replaced by `is-whitespace` or `is-not-whitespace`
- `is-color-enabled` replaced by `get-terminal-color-support --quiet`
- `is-empty-ls` replaced by `is-empty-directory`
- `is-empty-string` replaced by `is-whitespace`
- `is-nonempty-string` replaced by `is-not-whitespace`
- `get-line-count` replaced by `echo-count-lines`
- `rm-junk` replaced by `fs-trim --cache`
- `rm-modules` replaced by `fs-trim --module`

deprecated beta commands:

- `echo-if-empty` replaced by `echo-with-whitespace-fallback`
- `echo-if-path` replaced by `echo-if-present`
- `is-dir` replaced by `is-directory`
- `is-either` replaced by `is-needle`
- `is-equal` replaced by `is-needle`
- `is-exec` replaced by `is-executable`
- `is-neither` replaced by `! is-needle`
- `rm-svn` replaced by  `find ...`
- `rm-sync` replaced by `find ...`

promoted from beta commands:

- `echo-if-path` promoted to `echo-if-present`
- `echo-last-line`
- `echo-values`
- `is-empty-directory`
- `is-even`
- `is-file`

demoted to beta commands:

- `echo-affirmative`
- `find-directories`

beta command updates:

- add `echo-if-directory`
- add `echo-if-executable`
- add `echo-truncate-lines`, however you probably just want `head -n <lines>`
- add `echo-with-empty-fallback`, probably needs to be renamed to `echo-with-zero-length-fallback`
- add `echo-with-whitespace-fallback`, renamed from `echo-if-empty`
- `eject-all`: use `fs-structure`
- `eval-on-empty-stdin`: rewrote for modern conventions, and support `grep` fallback for when `ifne` is not present
- `eval-on-not-empty-stdin`: rewrote for modern conventions
- `get-codec`: fix tests
- `is-suffix`: use new `bash.bash` helpers
- `pdf-decrypt`: rewrote for modern conventions
- `researchgate-rename`: use `echo-regexp` instead of `python-regex`, and fix `is-same` call
- `rm-vmware` renamed to `rm-vmware-fusion`
- `setup-server`: updated for new `is-*` fs commands

merges:

- 23000e5a59cb14f8a8bb7766bd791a564300007e
- a2a32e5c80bc1eeb5ee11a4c5825d35e619c1c0b
- 4359e7eee5ed05ce5f6089234fed6df6076a698e
- d1161c861709bd15796e3b46cffccc16cd55c86f
- c16bc4b0a2139a9e4606d922ee79f489cc152b36
- c22757df3e0bbd754734090607b3891b1655571e
- b2c587d84cfa83391b95c5d7ae76821d3a055ffe
- 6fd530126081ffe813f54364191f4a9af016c0a3
- fefefabb595a56c18d11535e647001c6277ed9da
- 0a99c0ac5b46202fc99e5f296f64887e600e5939
- 86fb00669dd144ad7708f699b3a0e0a928878691
- fd59406f196a83d52557210a18043ad00cca6020
- 68122cf23299d3e66a7615de26b46a05de86ba38
- 3ffd048f80d23183ae3239716ee9b56cb224df22
- 1fc467ccb380880adc6774a859cd44e671fa5adb
- 954ed17180f21923b1c03887b74a3a5194fa98ba
- b710c111a0e02ad96008798bda532a45d735c94e
- 85c30bccf067f079d5b4f6a7f025ec1e8a19ddff
- 574c6c48eea249835f9aaed409896c17858fdbf7
- 790b68620f93275fecf96732392f6a682a906670
- df03da8226babeee15bdcaa6dd0e207d2c5bce21
- b7df848e17d863e2b691f7d4198c2fffc9d7a375
- f9bacd0bdd557d31b82e681e7de321a65709a80b
- 2cbeae0fb55bd91704cc007dd7ad94e292a08035
- d88ed97aac7983fa35e97c346839f3925f8d601c
- 390120595a2309760e9ba9d4d8252d0760488e12
- e99ff8c5d3ab775caf3cedeb7a39862dd9160741
- a2942477d32d200db6aab1e191b2b85399a7e04f
- 2c09a07655caf7a4f9a3cdab10d7a048e08ed41b
- d6ea4b83122bd00fb81c79a5946fd03b722cbef6
- 164eac52e908bdb80e43c52e65c27abb4a106f50
- 0c16d51c9776a1a01edb949510e56b652d1bd028
- 08c4e30c2aa879a294c2f2b594109e1f41c4bc3d
- 8f8145b21742283f68b56b855d74e05bb46ed7b0
- eaebe6777ef493927fbe4be38c69baa4ab593b3c
- 09af2fc6d407fe93ec5c8a5ceebaddb657e57e13
- 3fe8afab05f774d9868c1dc38e81593433e3dc7b
- ae9f32fbe7864ebf160cbcbdb2d7f82399daf893
- 6eadbea508d5e5e0f035939017705cfa3331566e
- 87c652981fe481fc268b802e6a1ffc6546432cdc
- 84d0cf75953a5b87346ffdb1289a25959e5e83a2
- 0a47ca290283768b85078882b5f1deb398122075
- 247281fb8fde0d5e06f9a08604a3e61f868dfe04
- f1879304182f887b8831ee0439f4218de02d6185
- 195272677824b0fe34935c57ca3a577411be5f61
- 85e727be297f1a2922058bc4912c083c04f620d2
- a20ad1fcaadba210794f5f14070aecb555dfc538
- d2597939c65353de04f8e14b1dfe5141003a43ed
- cf7eceb87e93269ca872f51874ade737aa6621dc
- 93adb5a8e5506377be46906d71577c6f87ba0293
- 21e8fe9e7ea7b2857bbee896befa7666e83266dc
- fae6f84628dfbdb640fddf48e5066c69f736632b
- 7290028ed79510004feafce92762bc1880c0e8d6
- a6b6d5449afe7aac6d70ce72d261fcb431ecbb0f
- e3b8fc401ea1d4bfd64f2b359cc1483a9ec0d2cb
- 30ef3dba9e77d70ac8873e2bc2cbc2e0326ea769
- 399531a22ea1b0e0c5f174fcb0674c594762dfd5
- 23ab815197bb4ae713ff1fa29be9af2bb2f8427d
- e05fa9ca799e8f82b59f788e9309465e38b06cc4
- 2e799489e4a47abf55e2bf0ca5f499b8167a72c2
- 0031c62d1c7ba763ad39ab0b5d52e2bdc1899966
- eaefcec17d71bad93959333b1b345f2e514f31b4
- cc8af1248d3297701413f4dc60a769df053788d9
- 68d7252c9703075bbe54c1090846959676fcbf80
- 4bf4b833b957d5c57cd9fcb2f66f4665f3c7afd0
- ec9ef689ff63dfa9cff746bf953cedd3e60b03c5
- d5d12bc89422ded18c2914bfe5e39a538c9705c4
- 5c701ece2e20766becbb25f0bf965b9aa27ff3c8
- 5ce4b8b479a9bd652001b07b772ec58c79746d2c
- b446b97a748aae7ee79269ed9bb5e276ad755419
- 49f5ca1934815e46772dd78fedf4ebdddd04666d
- b073dec27abca24e4bb8b2433ef699967e296fff
- 22b480ead2b211a2546738d7d11cc551218214aa
- bd7019f83fcb4dbd36b1bef237f2f46ff60788e4
- bef3016c793a3b2702ad6c83b50a96bce01d4318
- 1f6b37be28bdb5b39cd3c1ddb2416f64d5f57800
- 39b921d047672bb189da469c5e37904d22ba5738
- 268afa429ca149e385958be1745b1eead01ea4fd
- b2c006565c3cf1211826068f3ce02ff8931366b6
- 964bf205905d4d7a060b38451a690a4342228198
- d7b48ff6df3036c48484e9e83a967650a787b5b3
- 84768ebea9a7a0ef58555407e44e0080f793f583
- 872f0b782bfbd9ae6eabaf90190391eebba36825
- cca8f9b1d6734edef1f60f1dd6ceeccc669af850
- cf1039588540784b4d6e3b52d5a217f2427eb6ee
- 365c1895b71b35c35dd288902334eded5ab54ce9
- 3aa038f778c264b94ff33238b1e335da6320099e
- 664120d0fd7d4a10ee9d282dacf856b6196bdc13
- 89c8d2948781c3f3a290a34c7fc06bf65310618c
- 6d494c7f722eca7715c769d23360924a4347df4d
- 0e4cf5bc11e0f0d7763194533fb2978ddf282898
- f9723a93be5e192e27aebe898257b6be17f9b193
- 995533ae765984b807fa4452abc9c219cdd46e88
- 4f1d1b7d2055317e5c0bc7c9c31525ba7b6b2764
- a489debb703c47a6a0a45a9c662ef4a530d2f1a1
- d80be4bd7720e9e04ca793d69cef4de8719adc00
- 41a47e798ab55884151eb57d41be4048791d6a76
- bbf7e1736c1e735f711dcb6aaa3c949b14efd25e
- b06ce34ba9e4abac41889d83314cb929771e9caa
- c4d9ac64970b6e849571381b5b5c26b66832bec9
- cfec77cccc42bfe8791bcc914df2218d7d442eec
- af962c507ed637807897962bd8ab7f5bc9ac8fa1
- 55ee7e686510182a5786acfe748187db1272c354
- 88c9a33b38eb81ba425075c44b3c42c98123bb3c
- db0187dd4c437601386593ef2e6c607ff98b9548
- ad5887cdeb9b3d6fa73527d259a34e8368e92617
- 3db5aba4d4d26042638fe179536f36746c674adb
- 6ffa78c9a238a4b7e53f401914b8272d01d98e3f
- f8eb3221f57da6b09a82f115e1303c7707bf9e7d
- b624ee8b08f6f1666dfbcaf6059bce5630c2d153
- 839a82c631ec023f1c68618e032cdf61a51ddb67
- e0cd9a06b73fe91ee51c131ddf80ac6fad9b777c
- b96a30f7aac525fa8b89b4479c26d87f5e4b389c
- 25fa014532faea1646e99091ef2229ce69adae5f
- ccda6c844d80fad20b2900957a2f545e1f17ee40
- ef8a3877e0a665c6a05e833db38274aeda5dc054
- 75a30d7b49ab0f66f911670af43a6c9da7cfb4ea
- 3bf0e557e9d41253bff4a31d00c69fac4a190fc7
- c2f2ca90bd313b30d2132f6e6beeb9d5974f865f
- 4627b754b278f3f47759a7e93ff5bf270304050f

### CODE REVIEW

File: commands/get-terminal-theme [hunk 1]
```diff
@@ -112,9 +112,9 @@ function get_terminal_theme() (
 			# default;default — Foreground color code default and background color code default.
 			# ANSI colors are 0-15, in which 7 and 15 are light grey, and 9-15 are light colors.
 			local bg="$COLORFGBG"
-			if test -n "$bg"; then
+			if [[ -n $bg ]]; then
 				bg="${bg##*;}" # trim everything prior to the last ;
-				if test "$bg" = 'default' || test "$bg" -le 6 -o "$bg" -eq 8; then
+				if [[ $bg == 'default' ]] || (is-digit -- "$bg" && [[ $bg -le 6 || $bg -eq 8 ]]); then
```
Comment (balupton) [inline commands/get-terminal-theme:117]: the introduction of is-digit here is a sanity check to prevent weird behaviour

**Review by molleweide — COMMENTED** [review]
testing.

conventions.md seemed to show as draft PR even though we had finished the PR.

**Comment by balupton** [comment]
these conventions need to be updated before merge:

https://github.com/bevry/dorothy/blob/dev/square-brackets/docs/scripting/conventions.md?plain=1#L204-L208
